src/Elasticsearch/Metadata/Resource/Factory/ElasticsearchOperationProviderResourceMetadataCollectionFactory.php

```diff
@@ -0,0 +1,94 @@
+
+use ApiPlatform\Elasticsearch\State\ElasticsearchCollectionProvider;
+use ApiPlatform\Elasticsearch\State\ElasticsearchItemProvider;
+use ApiPlatform\Metadata\CollectionOperationInterface;
+use ApiPlatform\Metadata\GraphQl\Operation as GraphQlOperation;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\Metadata\Operations;
+use ApiPlatform\Metadata\PersistenceMeansInterface;
+use ApiPlatform\Metadata\Resource\Factory\ResourceMetadataCollectionFactoryInterface;
+use ApiPlatform\Metadata\Resource\ResourceMetadataCollection;
+
+/**
+ * @internal
+ */
+class ElasticsearchOperationProviderResourceMetadataCollectionFactory implements ResourceMetadataCollectionFactoryInterface
+{
+    public function __construct(
+        private readonly ResourceMetadataCollectionFactoryInterface $decorated,
+    ) {
+    }
+
+    public function create(string $resourceClass): ResourceMetadataCollection
+    {
+        $resourceMetadataCollection = $this->decorated->create($resourceClass);
+
+        foreach ($resourceMetadataCollection as $i => $resourceMetadata) {
+            if ($operations = $resourceMetadata->getOperations()) {
+                $resourceMetadata = $resourceMetadata->withOperations(self::configureOperations($operations));
+            }
+
+            // $graphqlOperations and $operations have not same type, we cannot combine function
+            if ($graphQlOperations = $resourceMetadata->getGraphQlOperations()) {
+                $resourceMetadata = $resourceMetadata->withGraphQlOperations(self::configureOperations($graphQlOperations));
+            }
+
+            $resourceMetadataCollection[$i] = $resourceMetadata;
+        }
+
+        return $resourceMetadataCollection;
+    }
+
+    /**
+     * @template  T of Operations|array<GraphQlOperation>
+     *
+     * @phpstan-param  T $operations
+     *
+     * @phpstan-return T
+     */
+    private static function configureOperations(Operations|array $operations): Operations|array
+    {
+        foreach ($operations as $operationName => $operation) {
+            $configuredOperation = self::configureOperation($operation);
+            if ($configuredOperation === $operation) {
+                continue;
+            }
+            if ($operations instanceof Operations) {
+                $operations->add($operationName, $configuredOperation);
+            } else {
+                $operations[$operationName] = $configuredOperation;
+            }
+        }
+
+        return $operations;
+    }
+
+    private static function configureOperation(Operation $operation): Operation
+    {
+        if (null !== $operation->getProvider()) {
+            return $operation;
+        }
+        if (!$operation->getPersistenceMeans() instanceof PersistenceMeansInterface) {
+            return $operation;
+        }
+        if (false === $operation->getElasticsearch()) {
+            throw new \LogicException('You cannot set $elasticsearch to false while configuring an elasticsearch document');
```

Comment:
this is legacy code and should not be called here

Reply:
Do you think this exception should be thrown in `ApiResource` and `Operation`’s constructors?

Reply:
only in the MetadataFactory that uses `getElasticsearch`